streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.kstream.Branched;
+import org.apache.kafka.streams.kstream.BranchedKStream;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.apache.kafka.streams.kstream.internals.graph.ProcessorGraphNode;
+import org.apache.kafka.streams.kstream.internals.graph.ProcessorParameters;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BranchedKStreamImpl<K, V> implements BranchedKStream<K, V> {
+
+    private static final String BRANCH_NAME = "KSTREAM-BRANCH-";
+
+    private final KStreamImpl<K, V> source;
+    private final boolean repartitionRequired;
+    private final String splitterName;
+    private final Map<String, KStream<K, V>> result = new HashMap<>();
+
+    private final List<Predicate<? super K, ? super V>> predicates = new ArrayList<>();
+    private final List<String> childNames = new ArrayList<>();
+    private final ProcessorGraphNode<K, V> splitterNode;
+
+    BranchedKStreamImpl(final KStreamImpl<K, V> source, final boolean repartitionRequired, final NamedInternal named) {
+        this.source = source;
+        this.repartitionRequired = repartitionRequired;
+        this.splitterName = named.orElseGenerateWithPrefix(source.builder, BRANCH_NAME);
+
+        // predicates and childNames are passed by reference so when the user adds a branch they get added to

```

Comment:
I am wondering if it might be better to move this code into a `build` method that would be called within `defaultBranch()` / `noDefaultBranch()` ?

The pattern to pass in empty list that we modify later seems undesirable, and we should first build the list, and than pass them in -- otherwise, we make assumptions how `ProcessorParameters` and `ProcessorGraphNode` might be implemented what we should avoid.

Reply:
I clearly remember that something made me to write it this way, but I have to recall...

Reply:
Would love to learn about it. -- In general, it's easier to follow the same pattern throughout the code base. It easier to reason about the code that way, and also easier for people to learn the code base.

Reply:
Just saw your other comment: https://github.com/apache/kafka/pull/9107#issuecomment-751261181